core/src/saros/net/xmpp/contact/XMPPContactsService.java

```diff
@@ -0,0 +1,415 @@
+    }
+  }
+
+  private void contactsUpdated(Collection<String> addresses) {
+    if (roster == null) return;
+
+    for (String address : addresses) {
+      XMPPContact contact = contacts.get(address);
+      if (contact == null) {
+        log.error("Should not happen: Contact for " + address + " not found!");
+        continue;
+      }
+
+      RosterEntry entry = roster.getEntry(address);
+      if (entry == null) {
+        log.error("Should not happen: Contact found, but RosterEntry is missing for: " + address);
+        continue;
+      }
+
+      if (handlePendingSubscription(contact, entry)) notifyListeners(contact, UpdateType.STATUS);
+
+      if (contacts.setContactGroups(contact, entry.getGroups()))
+        notifyListeners(contact, UpdateType.GROUP_MAPPING);
+
+      if (contact.setNickname(entry.getName()))
+        notifyListeners(contact, UpdateType.NICKNAME_CHANGED);
+    }
+  }
+
+  private void contactChangedPresence(Presence presence) {
+    if (roster == null) return;
+
+    String address = presence.getFrom();
+    XMPPContact contact = contacts.get(address);
+    if (contact == null) {
+      log.error("Contact " + address + " for presence update not found!");
+      return;
+    }
+
+    JID fqJid = new JID(address);
+    if (presence.isAvailable()) {
+      handleContactResourceAvailable(contact, fqJid, presence);
+    } else {
+      handleContactResourceUnavailable(contact, fqJid);
+    }
+  }
+
+  private void contactCanceledSubscription(JID jid) {
+    XMPPContact contact = contacts.get(jid.getRAW());
+    if (contact == null) {
+      log.debug("Strange behaviour: no contact found for canceled subscription from " + jid);
+      return;
+    }
+
+    contact.removeResources();
+    contact.setBaseStatus(ContactStatus.TYPE_SUBSCRIPTION_CANCELED);
+    notifyListeners(contact, UpdateType.STATUS);
+  }
+
+  private boolean handlePendingSubscription(XMPPContact contact, RosterEntry entry) {
+    if (entry.getStatus() == ItemStatus.SUBSCRIPTION_PENDING)
+      return contact.setBaseStatus(ContactStatus.TYPE_SUBSCRIPTION_PENDING);
+    return false;
+  }
+
+  private void handleContactResourceAvailable(XMPPContact contact, JID fqJid, Presence presence) {
+    contact.setSubscribed();
+
+    ContactStatus status;
+    if (presence.isAway()) {
+      status = ContactStatus.newInstance(Type.AWAY, presence.getMode().toString());
+    } else {
+      status = ContactStatus.TYPE_AVAILABLE;
+    }
+    if (contact.setResourceStatus(fqJid, status)) notifyListeners(contact, UpdateType.STATUS);
+
+    discoveryService.querySarosSupport(
+        fqJid.getRAW(), handleQuerySarosSupportResult(fqJid, contact));
+  }
+
+  private void handleContactResourceUnavailable(XMPPContact contact, JID fqJid) {
+    discoveryService.removePresence(fqJid.getRAW());
+    if (contact.removeResource(fqJid)) notifyListeners(contact, UpdateType.STATUS);
+  }
+
+  /**
+   * Function to be called after a positive query.
```

Comment:
I do not think it is a good design to return a Consumer here. All other handleXYZ do something. Calling this method does nothing unless you invoke the specific method on the Consumer object that this method returns.